go/arrow/flight/flightsql/driver.go

```diff
@@ -0,0 +1,772 @@
+	var params []driver.NamedValue
+	for i, arg := range args {
+		params = append(params, driver.NamedValue{
+			Name:    fmt.Sprintf("arg_%d", i),
+			Ordinal: i,
+			Value:   arg,
+		})
+	}
+
+	return s.ExecContext(context.Background(), params)
+}
+
+// ExecContext executes a query that doesn't return rows, such as an INSERT or UPDATE.
+func (s *Stmt) ExecContext(ctx context.Context, args []driver.NamedValue) (driver.Result, error) {
+	if err := s.setParameters(args); err != nil {
+		return nil, err
+	}
+
+	if _, set := ctx.Deadline(); !set && s.timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, s.timeout)
+		defer cancel()
+	}
+
+	n, err := s.stmt.ExecuteUpdate(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// FIXME: For now we ignore the number of affected records as it seems like
+	// the returned value is always one.
+	_ = n
+
+	return &Result{affected: -1, lastinsert: -1}, nil
+}
+
+// Query executes a query that may return rows, such as a SELECT.
+func (s *Stmt) Query(args []driver.Value) (driver.Rows, error) {
+	var params []driver.NamedValue
+	for i, arg := range args {
+		params = append(params, driver.NamedValue{
+			Name:    fmt.Sprintf("arg_%d", i),
+			Ordinal: i,
+			Value:   arg,
+		})
+	}
+
+	return s.QueryContext(context.Background(), params)
+}
+
+// QueryContext executes a query that may return rows, such as a SELECT.
+func (s *Stmt) QueryContext(ctx context.Context, args []driver.NamedValue) (driver.Rows, error) {
+	if err := s.setParameters(args); err != nil {
+		return nil, err
+	}
+
+	if _, set := ctx.Deadline(); !set && s.timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, s.timeout)
+		defer cancel()
+	}
+
+	info, err := s.stmt.Execute(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	rows := Rows{}
+	for _, endpoint := range info.Endpoint {
+		reader, err := s.client.DoGet(ctx, endpoint.GetTicket())
+		if err != nil {
+			return nil, fmt.Errorf("getting ticket failed: %w", err)
+		}
+		record, err := reader.Read()
+		if err != nil {
+			return nil, fmt.Errorf("reading record failed: %w", err)
+		}
+
+		if rows.schema == nil {
+			rows.schema = record.Schema()
+		}
+		if !rows.schema.Equal(record.Schema()) {
+			return nil, fmt.Errorf("mixed schemas %w", ErrNotSupported)
+		}
+		rows.records = append(rows.records, record)
+	}
+
+	return &rows, nil
+}
+
+func (s *Stmt) setParameters(args []driver.NamedValue) error {
+	if len(args) == 0 {
+		s.stmt.SetParameters(nil)
+		return nil
+	}
+
+	var fields []arrow.Field
+	sort.SliceStable(args, func(i, j int) bool {
+		return args[i].Ordinal < args[j].Ordinal
+	})
+
+	for _, arg := range args {
+		dt, err := toArrowDataType(arg.Value)
+		if err != nil {
+			return fmt.Errorf("schema: %w", err)
+		}
+		fields = append(fields, arrow.Field{
+			Name:     arg.Name,
+			Type:     dt,
+			Nullable: true,
+		})
+	}
+
+	schema := s.stmt.ParameterSchema()
```

Comment:
if your `toArrowDataType` auto-conversion checks don't match the types in the ParameterSchema, you could end up with a panic or other failure. If `s.stmt.ParameterSchema()` is non-nil, you should use that for converting the arguments rather than inferring the data types.

Reply:
Currently I have nothing to test this against as the example server will not return a parameter schema. Do you have a way to check my (to be written) code will handle the schema correctly?

Reply:
It should be pretty easy to create a sample testing server to utilize for the tests by following the examples and have that return an artificial schema to test with.

Reply:
Looking at that code again, I'm using the schema if given and otherwise create one from the given values. I changed the code a bit to make this more clear, but I don't see what you mean in your initial comment. Can you elaborate a bit @zeroshade?

Regarding creating of a test-server, this is way beyond my current capabilities as it involves parsing the SQL statement itself for adding it to the SQLite mock _or_ writing a complete mockup FlightSQL server...

Reply:
For my initial comment, I was referring to the fact that when you attempt to create a new record, if the schema you pass doesn't match the types of the slice of `arrow.Array` you use, then `NewRecord` will panic. So if the conversions you were doing in `toArrowDataType` don't match up with the schema returned from `ParameterSchema` you'd end up getting a panic when you try to create the record due to the mismatch. that's all.

> Regarding creating of a test-server, this is way beyond my current capabilities as it involves parsing the SQL statement itself for adding it to the SQLite mock or writing a complete mockup FlightSQL server...

You don't need a *complete* mockup flightsql server, you only need to implement the necessary methods that would return the ParameterSchema and you can hardcode the schema it returns. For example take a look at the mocked out server used for the ADBC timeout tests: https://github.com/apache/arrow-adbc/blob/main/go/adbc/driver/flightsql/flightsql_adbc_test.go#L709 

it's pretty straightforward I think.